### PR TITLE
[P2P] Improve RainTree counter mechanism in testing framework

### LINKED ISSUE #277: [TECHDEBT] [P2P] Raintree test suite improvements: Double check expected calls counting
## Objective

Tend for this TODO left in the code:

```go
./p2p/module_raintree_test.go:// INVESTIGATE(olshansky): Double check that how the expected calls are counted is accurate per the
```

## Origin Document

Looking at the configuration for a single RainTree test:

Screen Shot 2022-10-15 at 4 49 52 PM

Along with the visuals and the configs for a single test:

Screen Shot 2022-10-15 at 4 38 07 PM

It is not clear what these are supposed to mean in the context of gossip.

## Goals

- [ ] Make the RainTree testing framework counters more intuitive in terms of reads/writes from/to the network
- [ ] Add some clarifying documentation to the RainTree testing framework/modules

## Deliverable

A PR that includes the following
- [ ] Changes to how the RainTree testing framework counters are managed and validated
- [ ] Supporting documentation on what the counters mean and why

## Non-goals / Non-deliverables

- Modify Raintree implementation
- Code cleanup
- Code refactoring
- Add the tooling/utilities to gain visibility into RainTree
- Add the tooling/utilities to generate RainTree unit tests

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

## Testing Methodology

- **Task specific tests**: `make test_p2p`
- **All tests**: `make test_all`
- **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @deblasis 
**Co-Owners**: @Olshansk

### PR DESCRIPTION
**Pull request by Olshansk**
## Description

Improve the clarity of the counters mean (i.e. reads / writes to from / to the network) in the RainTree testing framework.

## Issue

Fixes #277

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other 

## List of changes

- Add a telemetry `send` event within the context RainTree network module that is triggered during network writes
- Change the RainTree testing framework counting method to simulate real reads/writes from the network
- Improve documentation related to the RainTree testing framework & how the counters are computed

## Testing

- [x] `make test_p2p`
- [x] `make develop_test`
- [ ] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: p2p/utils_test.go
Comment (andrewnguyen22) [inline p2p/utils_test.go:25]: We are currently trying to remove the test_artifacts package and fully encapsulate tests to their respective modules. I don't think this `TODO` is appropriate

File: p2p/module_raintree_test.go
Comment (andrewnguyen22) [inline p2p/module_raintree_test.go:148]: It's very strange to me that these numbers continue to change... This is the third time now. I think that this emphasizes further the need to move the `source of truth` code to this repository in order for formal verification. At this time, we are simply relying on these hard coded constants - which clearly are not stable

File: runtime/test_artifacts/generator.go
Comment (andrewnguyen22) [inline runtime/test_artifacts/generator.go:165]: TODO revert or DISCUSS how we are importing `generator.go` code in order to satisfy previously encapsulated tests. If we decide to have a shared `testing utils` package - that's fine, but the `generator` has a specific use case to create the build files and I don't think this is an appropriate import

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]

File: p2p/module_raintree_test.go
Comment (Olshansk) [reply p2p/module_raintree_test.go:148]: > It's very strange to me that these numbers continue to change... This is the third time now. 

It is the second time they're changing. It's the same change as in #179 which was never finished & merged in.

I decided to reduce the scope of that PR into several smaller work tasks per the comment here: [the testing frame](https://github.com/pokt-network/pocket/pull/179#issuecomment-1281126451).

I believe my first approach was naive, but this second one makes more sense (with comments in some tests to help explain it):

```
func TestRainTreeNetworkCompleteFourNodes(t *testing.T) {
	// Test configurations (visualization retrieved from simulator)
	// 	                val_1
	// 	  ┌───────────────┴────┬─────────────────┐
	//  val_2                val_1             val_3
	//    └───────┐            └───────┐         └───────┐
	// 		    val_3                val_2             val_4
	originatorNode := validatorId(1)
	var expectedCalls = TestNetworkSimulationConfig{
		originatorNode: {0, 3}, // val_1 does 3 network writes (two to val_2 and 1 to val_3)
		validatorId(2): {2, 1}, // val_2 does 2 network reads (both from val_1) and 1 network write (to val_3)
		validatorId(3): {2, 1}, // val_2 does 2 network reads (from val_1 and val_2) and 1 network write (to val_4)
		validatorId(4): {1, 0}, // val_2 does 1 network read (from val_3)
	}
	testRainTreeCalls(t, originatorNode, expectedCalls)
}
```

File: runtime/test_artifacts/generator.go
Comment (Olshansk) [reply runtime/test_artifacts/generator.go:165]: I fixed the encapsulation logic. Lmk if you want me to remove this change too though since it is a "fly by side change".

I have this comment in the `p2p/README` but might be worth discussing (or scoping into a specific piece of work):

```
- **Variable Coupling**:There is an implicit coupling between `validatorId`, `serviceUrl` and `genericParam` that requires understanding of the codebase. Reach out to @olshansk or @andrewnguyen22 for clarity on this.
```

File: p2p/module_raintree_test.go
Comment (andrewnguyen22) [reply p2p/module_raintree_test.go:148]: > It is the second time they're changing. It's the same change as in https://github.com/pokt-network/pocket/pull/179 which was never finished & merged in.

My mistake - sorry



Cool makes sense - can you link me the follow up issues that satisfy:
> I think that this emphasizes further the need to move the source of truth code to this repository in order for formal verification
I want to make sure this is well documented and the intention is clear before unblocking this review

File: p2p/README.md
Comment (andrewnguyen22) [inline p2p/README.md]: Can you explain to me the reasoning for removing this?

File: p2p/utils_test.go
Comment (andrewnguyen22) [inline p2p/utils_test.go:90]: A little more detail would be nice

File: runtime/test_artifacts/generator.go
Comment (andrewnguyen22) [inline runtime/test_artifacts/generator.go:165]: Use this function in `validatorId()`

Remove duplicate `ServiceUrlFormat` in favor of `serviceUrlFormat`

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]

File: p2p/README.md
Comment (Olshansk) [reply p2p/README.md]: I believe I sent the original PR before it was review for review. Those were just rough "talking notes" from our meeting.

The real delta (relative to main) is this:

Screen Shot 2022-10-17 at 11 02 20 AM

File: runtime/test_artifacts/generator.go
Comment (Olshansk) [reply runtime/test_artifacts/generator.go:165]: Should I revert it to what I had before? I removed the coupling of the unit test helpers in p2p on runtime based on the previous comment:

Screen Shot 2022-10-17 at 11 05 31 AM

File: p2p/module_raintree_test.go
Comment (Olshansk) [reply p2p/module_raintree_test.go:148]: @jessicadaugherty Can you linkn to the ticket for (1) from https://github.com/pokt-network/pocket/pull/179#issuecomment-1281133270 when you have a chance?

**Comment by Olshansk** [comment]
@andrewnguyen22 PTAL at my responses again when you have a chance

File: p2p/module_raintree_test.go
Comment (jessicadaugherty) [reply p2p/module_raintree_test.go:148]: Yes will include as part of origin doc. Thanks @Olshansk!

**Review by deblasis — COMMENTED** [review]
Definitely the comments and the additional context in the README.md help a lot following what's going on. 

The tests follow what the reader might expect (arrange -> act -> assert).

Assuming that there are not going to be "black boxes" and that the reader can have a way of generating new test cases and/or "ascii diagrams", I think this is good enough as a foundational piece of work we can iterate on. 

I am not sure I understand why the counter values have changed (I glanced at #179 and the discussion below) but I gather that the previous implementation was simply incorrect, and the current one is correct.

**Comment by Olshansk** [comment]
> Definitely the comments and the additional context in the README.md help a lot following what's going on.
> 
> The tests follow what the reader might expect (arrange -> act -> assert).

Any more required changes from your end to merge this in?

> Assuming that there are not going to be "black boxes" and that the reader can have a way of generating new test cases and/or "ascii diagrams", I think this is good enough as a foundational piece of work we can iterate on.

@jessicadaugherty Will help scope that work out into another task.

> I am not sure I understand why the counter values have changed (I glanced at #179 and the discussion below) but I gather that the previous implementation was simply incorrect, and the current one is correct.

I do not think my initial approach was clear and intuitive, hence the TODO in #277 while I kept thinking about it.

This is the 1st change (i.e. 2nd attempt) and the counters and I don't think it'll change anymore (unless there's a change to how RainTree works). Again, IMO, this makes a lot of sense but asking for a review as well:

```
func TestRainTreeNetworkCompleteFourNodes(t *testing.T) {
	// Test configurations (visualization retrieved from simulator)
	// 	                val_1
	// 	  ┌───────────────┴────┬─────────────────┐
	//  val_2                val_1             val_3
	//    └───────┐            └───────┐         └───────┐
	// 		    val_3                val_2             val_4
	originatorNode := validatorId(1)
	var expectedCalls = TestNetworkSimulationConfig{
		originatorNode: {0, 3}, // val_1 does 3 network writes (two to val_2 and 1 to val_3)
		validatorId(2): {2, 1}, // val_2 does 2 network reads (both from val_1) and 1 network write (to val_3)
		validatorId(3): {2, 1}, // val_2 does 2 network reads (from val_1 and val_2) and 1 network write (to val_4)
		validatorId(4): {1, 0}, // val_2 does 1 network read (from val_3)
	}
	testRainTreeCalls(t, originatorNode, expectedCalls)
}
```

File: runtime/test_artifacts/generator.go
Comment (andrewnguyen22) [reply runtime/test_artifacts/generator.go:165]: Nope - my comment is wrong. Good call out - we should not import test_artifacts here

File: p2p/module_raintree_test.go
Comment (andrewnguyen22) [reply p2p/module_raintree_test.go:148]: Sorry if I'm out of sync here: 

> Adding the Jupyter notebook to generate tests

Is this Ticket 1? I'm not sure if that satisfies my ask.

Waiting for this linking to approve

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]

File: p2p/module_raintree_test.go
Comment (Olshansk) [reply p2p/module_raintree_test.go:150]: I created #314 as follow up work specific to the tooling used to generate the tests.

Also added 2 TODOs in the code referencing it.

**Comment by Olshansk** [comment]
@andrewnguyen22 Let me know if the creation, details in, and reference to #314 is enough to unblock this PR.

**Review by andrewnguyen22 — APPROVED** [review]

**Pull request merged**
